libsnark/zk_proof_systems/plonk/tests/test_plonk.cpp

```diff
@@ -1095,6 +1095,142 @@ template<typename ppT> void test_plonk_gates_matrix_transpose()
     ASSERT_EQ(gates_matrix_transpose, example.gates_matrix_transpose);
 }
 
+// We test the example circuit y^2 = x mod r where x=49 is a public
+// input and y is the witness. We prove that we know y=7 such that y
+// is the quadratic residue of x modulo the modulus r of the scalar
+// field Fr of the given elliptic curve. The circuit is represented by
+// one public input (PI) and one multiplication gate. According to the
+// Plonk arithmetization rules, each gate (including the PI) is
+// represented as:
+//
+// qL a + qR b + qO c + qM ab + qC = 0
+//
+// where qL,qR,qO,qM,qC are the selector polynomials and
+// a,b,c are respectively the vectors of left inputs, right inputs
+// and outputs to the gates.
+//
+// Using the above representation, the PI and the multiplication
+// gates are given resp. as
+//
+//  PI: 1 a1 + 0 b1 +   0  c1 + 0 a1 b1 + 0 = 49
+// MUL: 0 a1 + 0 b1 + (-1) c1 + 1 a1 b1 + 0 = 0
+//
+// where
+//
+// a   = (a1, a2) = (x, y)
+// b   = (b1, b2) = (1, y)
+// c   = (c1, c2) = (x, x)
+// qL = (1,  0)
+// qR = (0,  0)
+// qO = (0, -1)
+// qM = (0,  1)
+// qC = (0,  0)
+//
+// The selector polynomials as given above define the following gates
+// matrix:
+//
+// qL qR qO qM qC
+//  1  0  0  0  0
+//  0  0 -1  1  0
+//
+// and the permutation expressing the copy-constraints is (second
+// line below):
+//
+// (a1 a2 b1 b2 c1 c2)
+// (c1 b2 b1 a2 c2 a1)
+//
+// reflecting the fact that a1=c1=c2=x and a2=b2=y
+template<typename ppT, class transcript_hasher>
+void test_plonk_prepare_gates_matrix()
+{
+    using Field = libff::Fr<ppT>;
+
+    // 0 Arithmetization of test circuit y^2 = x mod r
+
+    // The number of gates is 2: one public input (PI) gate and one
+    // multiplication gate. Tis nummer is also conveniently a power of
+    // 2 (needed for the FFT/iFFT)
+    const size_t num_gates = 2;
+    // The tested circuit has 1 public input
+    const size_t num_public_inputs = 1;
+    std::vector<std::vector<Field>> gates_matrix =
+        plonk_prepare_gates_matrix<ppT>(num_public_inputs);
+    ASSERT_EQ(gates_matrix.size(), num_public_inputs);
+    // Add the PI gate/s
+    const std::vector<Field> PI_gate{1, 0, 0, 0, 0};
+    for (size_t i = 0; i < num_public_inputs; ++i) {
```

Comment:
This seems to be a generic check for the `plonk_prepare_gates_matrix` function, but here we only ever set `num_public_inputs` to 1.  How about we do this check as a separate unit tests, and run this generic check for a few different values of `num_public_inputs`.  Then we can remove this code from this test (which is already pretty big).

The point of creating a simple circuit was not necessarily to replace the simple unit test, but to find any more low-hanging fruit in terms of convenience functions.  For example, in a discussion the other day, I mentioend that we don't want the caller to have to generate the set of PI indices.

Reply:
Fixed to the best of my understanding, but let's also briefly discuss offline.